src/core/ensemble_scheduler.cc

```diff
@@ -1198,24 +1198,33 @@ EnsembleContext::ScheduleSteps(
 {
   for (auto& step : steps) {
     step->ctx_ = context;
+    bool should_schedule = false;
     {
       std::lock_guard<std::mutex> lock(context->mutex_);
 
       // Need to check the ensemble_status_ to ensure the FinishEnsemble()
       // is called only once.
       if (context->ensemble_status_.IsOk()) {
         context->request_tracker_->IncrementCounter();
-        context->ensemble_status_ = context->is_->InferAsync(step->request_);
-        if (!context->ensemble_status_.IsOk()) {
-          // The request is not sent to server properly, shouldn't expect its
-          // release function get called.
-          context->request_tracker_->DecrementCounter();
-          context->ensemble_status_ = context->FinishEnsemble();
-          break;
-        }
+        should_schedule = true;
+      }
+    }
+    if (should_schedule) {
+      // TODO: Move ownership of request out of ensemble step to avoid
+      // issues with double release on request on cache hits
```

Comment:
@GuanLuo for wording on this comment / WAR

Reply:
"On a successful call to InferAsync(), the ensemble step is managed by the response callback. When the response callback is invoked, the request must not be owned by step as the request should be transferred and managed by Triton core. In the case of cache hit, the request hasn't been transferred and can cause double-free, so moving the request ownership out of step here to avoid that."

Reply:
done